### DISCUSSION TITLE
Live preview of docs in VSCode

### DISCUSSION THREAD

**Discussion by samsrabin** [ESCOMP/CTSM#2093] [Show and tell]
I've been doing a bit of work on the docs and finally worked out a good way to preview them from within VSCode. It takes a bit of effort, but once it's working it's pretty nice.

![screenshot_6140](https://github.com/ESCOMP/CTSM/assets/10454527/beb1946c-d0f0-4ca2-8775-ac838ad2fc93)

# Instructions

The first step is to install the [reStructuredText extension for VSCode](https://marketplace.visualstudio.com/items?itemName=lextudio.restructuredtext). Their website has some [documentation](https://docs.restructuredtext.net/) that's helpful for understanding what it's useful for and how to set it up (note especially [Prerequisites](https://docs.restructuredtext.net/articles/prerequisites) and [Configuration](https://docs.restructuredtext.net/articles/configuration)). I'll describe my process here.

2. Set up a clean virtual environment for VSCode to use for this. I used `mamba` but it should work equally well with `conda`. _Instructions updated 2024-11-01._
```bash
mamba create --name vscode-restructuredtext
mamba activate vscode-restructuredtext
mamba install -c conda-forge python esbonio sphinx-rtd-theme sphinx-mdinclude
```

3. Set VSCode to use the new virtual environment as the Python interpreter: Open the Command Palette and search for `Python: Select Interpreter`.
4. In VSCode Workspace settings:
- Set `Esbonio › Server: Python Path` to the path you get when you do `which python` in your `vscode-restructuredtext` environment.
- Set `Esbonio › Sphinx: Conf Dir` to `${workspaceFolder}/doc/source`.
- Optional: Set `Esbonio › Sphinx: Build Dir` to `${workspaceFolder}/.vscode/_build_esbonio_sphinx/`. This will make it so that the generated HTML files etc. are ignored by Git.
- Optional: Uncheck `Esbonio › Sphinx: Force Full Build`.
- Optional: Set `Esbonio › Sphinx: Num Jobs` to 0. This will allow `sphinx-build` to automatically determine how many processes to use when generating the HTML.
5. I quit and reopened VSCode at this point, but I'm not sure that's necessary.
6. From the View menu, click Output, then in that panel's dropdown menu select Esbonio Language Server. This will show progress and messages from the generation process.
7. If it's not already going, click "esbonio" at the bottom right of the VSCode window (see my screenshot above). This should start the build process. The first time will take a while!
8. Once it's done, open a `.rst` file from the docs (e.g., `doc/source/tech_note/Methane/CLM50_Tech_Note_Methane.rst`). Open the Command Palette and do `reStructured Text: Open Preview to the Side`. This should show the [Tech Note](https://escomp.github.io/ctsm-docs/versions/master/html/tech_note/Methane/CLM50_Tech_Note_Methane.html) we all know and love!
9. Try making some change, like adding "Sam rocks" somewhere in the text. When you save, it should kick off a re-render.

# Tweaking the .rst files

There are some things that didn't work right until I actually edited the docs themselves.

Mathematical symbols and equations were mostly or all broken:
![screenshot_6141](https://github.com/ESCOMP/CTSM/assets/10454527/e4a95e98-0ae4-4b02-ab7f-72492732e8b4)

I solved this by deleting the `'sphinx.ext.imgmath',` line from `doc/source/conf.py`:
![screenshot_6143](https://github.com/ESCOMP/CTSM/assets/10454527/2426c47d-ccdf-4fb1-88cb-89839963c731)

Some equations still didn't render:
![screenshot_6142](https://github.com/ESCOMP/CTSM/assets/10454527/c0d1ed9e-1b45-4aee-928c-dc2ef0562fd8)

I solved this by deleting all instances of `\kern-\nulldelimiterspace` in all `.rst` files:
![screenshot_6144](https://github.com/ESCOMP/CTSM/assets/10454527/b024bc55-1d3a-4f67-852b-eb09ce358ab2)

# Caveats
Clicking links in the Live Preview opens my web browser to some unfindable page.

There are some warnings about failed references that I want to track down:
```
doc/source/lilac/obtaining-building-and-running/setting-ctsm-runtime-options.rst:107: WARNING: undefined label: 'creating-domain-files'
doc/source/lilac/obtaining-building-and-running/setting-ctsm-runtime-options.rst:111: WARNING: undefined label: 'creating-surface-datasets'
doc/source/lilac/obtaining-building-and-running/setting-ctsm-runtime-options.rst:137: WARNING: undefined label: 'customizing-a-case'
doc/source/tech_note/Introduction/CLM50_Tech_Note_Introduction.rst:52: WARNING: Failed to create a cross reference. Any number is not assigned: figure-mosart-conceptual-diagram
doc/source/tech_note/MOSART/CLM50_Tech_Note_MOSART.rst:28: WARNING: Failed to create a cross reference. Any number is not assigned: figure-mosart-conceptual-diagram
doc/source/users_guide/using-clm-tools/creating-input-for-surface-dataset-generation.rst:114: WARNING: undefined label: 'figure mksurfdatamap'
doc/source/users_guide/using-clm-tools/creating-surface-datasets.rst:19: WARNING: undefined label: 'figure global_domain'
doc/source/users_guide/using-clm-tools/creating-surface-datasets.rst:19: WARNING: undefined label: 'figure mksurfdatamap'
```

# Questions
- Would it be possible for us to move away from relying on `sphinx.ext.imgmath`? The base `sphinx` seems to include math rendering capabilities now.
- Was `\kern-\nulldelimiterspace` doing something important?
- Would anyone else be willing to try and set this up? I'd like to know if there are steps I forgot to add or other roadblocks.

**Comment by ekluzek** [comment]
@olyson do you know the answer to @samsrabin questions above?

**Comment by billsacks** [comment]
Cool, @samsrabin - thanks for writing this up!

I'm not sure about this:

> Would it be possible for us to move away from relying on sphinx.ext.imgmath?

But I do know that this line:

`imgmath_image_format = 'svg'`

was important to get better-looking equations in the rendered documentation. I'm not sure whether that relies on the sphinx.ext.imgmath package.

**Reply by samsrabin** [reply]
Thanks, Bill! I tried deleting that line just now, and indeed it broke all the equations. However, weirdly, it doesn't seem to rely on `sphinx.ext.imgmath`, as just deleting the latter works fine. I'm thinking maybe Esbonio doesn't support `sphinx.ext.imgmath` because the functionality, including `imgmath_image_format`, is now included in base Sphinx.

**Reply by samsrabin** [reply]
Actually, no, I was wrong—somehow I had put `sphinx.ext.imgmath` back in conf.py. Removing both that line and `imgmath_image_format` seems to work fine.

**Reply by billsacks** [reply]
If you want to experiment with removing the imgmath_image_format line, the important thing to check is the clarity of the equations in the generated html that we push up to https://escomp.github.io/ctsm-docs . Before that change, equations were generated as png files that had pretty poor quality. Changing to svg led to much crisper display of the equations - though a downside, if I remember correctly, was significantly longer documentation build times.

**Reply by samsrabin** [reply]
Good point. The quality looks fine to me; what do you think? Here's an [example .html file](https://www.dropbox.com/scl/fi/yyfnnt7agsrde77bf180o/CLM50_Tech_Note_Methane.html?rlkey=f1krxiuxmq2hweolpdbzna25f&dl=0).

**Reply by billsacks** [reply]
That looks good to me, too. Maybe the earlier issues are better now?
